src/odemis/driver/actuator.py

```diff
@@ -2179,3 +2185,631 @@ def terminate(self):
+
+    def _cancelCurrentMove(self, future):
+        """
+        Cancels the current move (both absolute or relative) or current reference
+        future (Future): the future to stop.
+        return (bool): True if it successfully cancelled (stopped) the move.
+        """
+        logging.debug("Cancelling current move...")
+        future._must_stop.set()  # tell the thread taking care of the move it's over
+        with future._moving_lock:
+            if future._state == FINISHED:
+                return False
+            # If future has and underlying move it should be cancelled
+            if future._running_subf:
+                future._running_subf.cancel()
+        return True
+
+
+class LinkedHeightFocus(model.Actuator):
+    """
+    This wrapper is the child of LinkedHeightActuator, it controls the lens stage Z axis and provide the means to convert it to
+    focus value and vice versa. It also shares the same executor for the movement functions with the parent.
+    """
+
+    def __init__(self, name, role, parent, dependencies, rng, executor, **kwargs):
+        """
+        :param name: (string)
+        :param role: (string)
+        :param parent (Component): the parent of this component, that will be in
+        .parent (the LinkedHeightActuator object)
+        :param dependencies (dict str -> actuator): axis name (in this
+        actuator) -> actuator to be used for this axis (Presumably the lens Z axis)
+        :param rng (tuple float, float): the focus range
+        """
+        # Get the underlying lens stage from the passed dependencies list
+        self._lensz = dependencies["lensz"]
+        # Set focus axis and make sure its range is not bigger than the underlying lens range
+        if rng[0] > rng[1]:
+            raise ValueError("Range left side should be less than right side.")
+        if (rng[1] - rng[0]) > (self._lensz.axes['z'].range[1] - self._lensz.axes['z'].range[0]):
+            raise ValueError("Focus range should be lower than the underlying lens range.")
+        self._range = rng
+        axes_def = {
+            "z": model.Axis(unit="m", range=rng),
+        }
+        # Get the shared executor from the parent
+        self._executor = executor
+        model.Actuator.__init__(self, name, role, parent=parent, dependencies=dependencies, axes=axes_def, **kwargs)
+
+        self.position = model.VigilantAttribute({}, unit="m", readonly=True)
+        # Position will be updated whenever the underlying lens stage is changed
+        self._lensz.position.subscribe(self._updatePosition, init=True)
+
+        # Check if underlying lens has FAV_POS_DEACTIVE metadata
+        if model.MD_FAV_POS_DEACTIVE not in self._lensz.getMetadata():
+            raise ValueError("The underlying lens stage is missing a FAV_POS_DEACTIVE metadata.")
+        # Check if underlying lens FAV_POS_DEACTIVE is not within its active range
+        lens_deactive = self._lensz.getMetadata()[model.MD_FAV_POS_DEACTIVE]['z']
```

Comment:
Preferably, call `.getMetadata()` only once, and reuse the return value. Or use a `try... except KeyError` to do the same.